{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def codon_dir(codon):\n",
    "    if codon in {\"TAA\",\"TAG\",\"TGA\"}:#stop codon for foward lane\n",
    "        return 1\n",
    "    elif codon in {\"TTA\",\"CTA\",\"TCA\"}:#stop codon for backward lane\n",
    "        return -1\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "filepath=\"/home/mitsuki/out/altorf/genome/fasta/GCF_000265365.1_ASM26536v1_chromosome.fna\"\n",
    "\n",
    "dct_lst=[]\n",
    "for seqRecord in SeqIO.parse(filepath, \"fasta\"):\n",
    "    for i in range(2,len(seqRecord)):\n",
    "        codon=seqRecord[i-2:i+1].seq\n",
    "        codonDir=codon_dir(str(codon))\n",
    "        \n",
    "        if codonDir!=0:\n",
    "            dct={}\n",
    "            dct[\"Chromosome\"]=seqRecord.name\n",
    "            dct[\"Start\"]=i-2\n",
    "            dct[\"End\"]=i+1\n",
    "            \n",
    "            lane_lst=[\"+1\",\"+2\",\"+3\",\"-1\",\"-2\",\"-3\"]\n",
    "            for lane in lane_lst:\n",
    "                dct[lane]=0\n",
    "            \n",
    "            for i in range(1,4):\n",
    "                for strand in ['+','-']:\n",
    "                    dct[strand+str(i)]=0                \n",
    "            \n",
    "            if codonDir==1:\n",
    "                dct[\"Feature\"]=str(codon)\n",
    "                dct['+'+str((i-2)%3+1)]=1\n",
    "                \n",
    "                idx=(i-2)%3+4\n",
    "            else:\n",
    "                dct[\"Feature\"]=str(codon.reverse_complement())\n",
    "                dct['-'+str((i-2)%3+1)]=1\n",
    "            dct_lst.append(dct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Feature</th>\n",
       "      <th>+1</th>\n",
       "      <th>+2</th>\n",
       "      <th>+3</th>\n",
       "      <th>-1</th>\n",
       "      <th>-2</th>\n",
       "      <th>-3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>43</td>\n",
       "      <td>46</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>46</td>\n",
       "      <td>49</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>52</td>\n",
       "      <td>55</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>71</td>\n",
       "      <td>74</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>95</td>\n",
       "      <td>98</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>103</td>\n",
       "      <td>106</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>133</td>\n",
       "      <td>136</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>206</td>\n",
       "      <td>209</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>230</td>\n",
       "      <td>233</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>253</td>\n",
       "      <td>256</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>266</td>\n",
       "      <td>269</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>290</td>\n",
       "      <td>293</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>314</td>\n",
       "      <td>317</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>328</td>\n",
       "      <td>331</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_018011.1</td>\n",
       "      <td>364</td>\n",
       "      <td>367</td>\n",
       "      <td>TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome  Start  End Feature  +1  +2  +3  -1  -2  -3\n",
       "0   NC_018011.1      5    8     TAA   0   0   0   0   1   0\n",
       "1   NC_018011.1      9   12     TGA   0   0   0   0   1   0\n",
       "2   NC_018011.1     20   23     TGA   0   0   0   0   1   0\n",
       "3   NC_018011.1     25   28     TGA   0   0   0   0   1   0\n",
       "4   NC_018011.1     43   46     TAA   0   0   0   0   1   0\n",
       "5   NC_018011.1     46   49     TGA   0   0   0   0   1   0\n",
       "6   NC_018011.1     51   54     TAA   0   0   0   0   1   0\n",
       "7   NC_018011.1     52   55     TAA   0   1   0   0   0   0\n",
       "8   NC_018011.1     71   74     TAG   0   1   0   0   0   0\n",
       "9   NC_018011.1     95   98     TGA   0   1   0   0   0   0\n",
       "10  NC_018011.1    103  106     TGA   0   0   0   0   1   0\n",
       "11  NC_018011.1    133  136     TGA   0   1   0   0   0   0\n",
       "12  NC_018011.1    206  209     TAG   0   0   0   0   1   0\n",
       "13  NC_018011.1    230  233     TAG   0   0   0   0   1   0\n",
       "14  NC_018011.1    253  256     TGA   0   0   0   0   1   0\n",
       "15  NC_018011.1    266  269     TAG   0   0   0   0   1   0\n",
       "16  NC_018011.1    290  293     TAG   0   0   0   0   1   0\n",
       "17  NC_018011.1    314  317     TGA   0   0   0   0   1   0\n",
       "18  NC_018011.1    328  331     TGA   0   0   0   0   1   0\n",
       "19  NC_018011.1    364  367     TGA   0   0   0   0   1   0"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df=pd.DataFrame(dct_lst)\n",
    "out_df=out_df[[\"Chromosome\",\"Start\",\"End\",\"Feature\"]+lane_lst]\n",
    "out_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3734239"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqRecords[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_018011.1'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqRecord.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "for i in range(1,4):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
